scripts/emulabconfig.py

```diff
@@ -71,69 +127,104 @@ def get_remote_scripts_path(self):
+                'mkdir -p $(dirname %s)/shm; ' % self.cluster.log_subdir +
+                'mkdir -p %s; ' % self.cluster.log_subdir +
+                'rm logs/latest; ' +
+                'ln -sf $(basename %s) logs/latest)' % self.cluster.log_subdir)
+
+    def fix_disk_permissions(self, server, parallel=False):
         log("Fixing disk permissions on %s" % server)
         disks = default_disk2.split(' ')[1].split(',')
         cmds = ['sudo chmod 777 %s' % disk for disk in disks]
-        ssh(server, '(%s)' % ';'.join(cmds))
+        if parallel:
+            pdsh('(%s)' % ';'.join(cmds))
+        else:
+            ssh(server, '(%s)' % ';'.join(cmds))
 
     def cluster_enter(self, cluster):
         self.cluster = cluster
-        log('== Connecting to Emulab via %s ==' % hosts[0][0])
-        for host in hosts:
-            hostName = host[0]
-            log('-- Preparing host ' + hostName)
-            self.send_code(hostName)
-            self.compile_code(hostName)
-            self.create_log_dir(hostName)
-            self.fix_disk_permissions(hostName)
+        log('== Connecting to Emulab via %s ==' % self.hosts[0][0])
+        if self.parallel:
+            self.kill_procs('all', parallel=self.parallel)
+            self.send_code('all', parallel=self.parallel)
+            self.compile_code('all', parallel=self.parallel)
+            self.create_log_dir('all', parallel=self.parallel)
+            self.fix_disk_permissions('all', parallel=self.parallel)
+        else:
+            for host in self.hosts:
+                hostName = host[0]
+                log('-- Preparing host ' + hostName)
+                self.kill_procs(hostName)
+                self.send_code(hostName)
+                self.compile_code(hostName)
+                self.create_log_dir(hostName)
+                self.fix_disk_permissions(hostName)
         log('== Emulab Cluster Configured ==')
+        log("If you are running clusterperf, it might take a while!")
 
-    def collect_logs(self, server):
+    def collect_logs(self, server, parallel=False):
         log('Collecting logs from host ' + server)
+        if parallel:
+            for host in self.hosts:
+                self.collect_logs(host[0])
         subprocess.check_call("rsync -ave ssh " +
                               "%s:%s/logs/ logs/> /dev/null" % (server,
                                                          self.get_remote_wd()),
                               shell=True, stdout=sys.stdout)
 
     def cluster_exit(self):
         log('== Emulab Cluster Tearing Down ==')
-        for host in hosts:
+        for host in self.hosts:
             hostName = host[0]
             self.collect_logs(hostName)
         log('== Emulab Cluster Torn Down ==')
         pass
 
-hooks = EmulabClusterHooks()
-
 local_scripts_path = os.path.dirname(os.path.abspath(__file__))
 top_path = os.path.abspath(local_scripts_path + '/..')
-
 obj_path = os.path.join(top_path, obj_dir)
-
```

Comment:
This all looks pretty good. I'd find a way to pull the "parallel/non-parallel" selection stuff into a few lines of code, though. Or, I'd even be ok with killing the non-parallel version. You could even just have a wrapper function that calls pdsh by default by can alternatively foreach ssh instead.